crates/storage/provider/src/providers/blockchain_provider.rs

```diff
@@ -4207,6 +3744,103 @@ mod tests {
         Ok(())
     }
 
+    macro_rules! test_by_block_range {
+        ($provider:expr, $database_blocks:expr, $in_memory_blocks:expr, [$(($method:ident, $data_extractor:expr)),* $(,)?]) => {{
+            let db_block_count = $database_blocks.len() as u64;
+            let in_mem_block_count = $in_memory_blocks.len() as u64;
+
+            let db_range = 0..=db_block_count - 1;
+            let in_mem_range = db_block_count..=(in_mem_block_count + db_range.end());
+
+            $(
+                // Retrieve the expected database data
+                let database_data =
+                    $database_blocks.iter().map(|b| $data_extractor(b)).collect::<Vec<_>>();
+                assert_eq!($provider.$method(db_range.clone())?, database_data);
+
+                // Retrieve the expected in-memory data
+                let in_memory_data =
+                    $in_memory_blocks.iter().map(|b| $data_extractor(b)).collect::<Vec<_>>();
+                assert_eq!($provider.$method(in_mem_range.clone())?, in_memory_data);
+
+                // Test partial in-memory range
+                assert_eq!(
+                    &$provider.$method(in_mem_range.start() + 1..=in_mem_range.end() - 1)?,
+                    &in_memory_data[1..in_memory_data.len() - 1]
+                );
+
+
+                // Test range that spans database and in-memory
+                assert_eq!(
+                    $provider.$method(in_mem_range.start() - 2..=in_mem_range.end() - 1)?,
+                    database_data[database_data.len() - 2..]
+                        .iter()
+                        .chain(&in_memory_data[..in_memory_data.len() - 1])
+                        .cloned()
+                        .collect::<Vec<_>>()
+                );
+                // Test range in in-memory to unbounded end
+                // assert_eq!($provider.$method(in_mem_range.start() + 1..)?, &in_memory_data[1..]);
+
+                // Test range that spans database and in-memory with unbounded end
+                // assert_eq!(
+                //     $provider.$method(in_mem_range.start() - 2..)?,
+                //     database_data[database_data.len() - 2..]
+                //         .iter()
+                //         .chain(&in_memory_data[..])
+                //         .cloned()
+                //         .collect::<Vec<_>>()
+                // );
+
+                // Test empty range
+                let start_block_num = u64::MAX;
+                let end_block_num = u64::MAX;
+                let result = $provider.$method(start_block_num..=end_block_num-1)?;
+                assert!(result.is_empty(), "No data should be found for an empty block range");
```

Comment:
Maybe here you don't want to test this behavior but I thought it might be even better to add below a test with a non-empty range but that does not correspond to anything valid. For example with `start_block_num = 1000` and `end_block_num = 5000` so as to check that if the range is not empty but does not contain a block then the result must also be empty.

Even if the behavior is close I say that to validate the behavior that the range is empty or that it is not empty but with no block in it (the behaviors of the functions can potentially be different in both cases).